### PR TITLE
Generalize DRS client (#2150)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author (before primary review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `HCA` label <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes <sub>or this PR does not need porting</sub> 
- [x] Added `[r]` prefix at start of commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles <sub>or there are no freebies in this PR</sub>
- [x] Added `chain` label <sub>or this PR is not the base of another PR</sub>
- [x] Made this PR a blocker of next PR in chain <sub>or this PR is not the base of another PR</sub>

Primary reviewer (before pushing merge commit)

- [ ] Checked `reindex` label and `[r]` prefix of commit title
- [ ] Rebased and squashed branch
- [ ] Sanity-checked history
- [ ] Build passes in `sandbox` <sub>or commented that sandbox will be skipped</sub>
- [ ] Reindexed `sandbox` <sub>or this PR does not require reindexing</sub>
- [ ] Added PR reference to merge commit

Primary reviewer (after pushing merge commit)

- [ ] Pushed merge commit to Github and Gitlab
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations <sub>or labeled as `no demo`</sub>
- [ ] Reindexed `dev` <sub>or this PR does not require reindexing</sub>

### CODE REVIEW

File: src/azul/drs.py [hunk 1]
```diff
@@ -254,36 +254,49 @@ def access_url(url):
     return {'url': url}
 
 
-class Client:
+@dataclass
+class ClientObject:
```
Comment (jessebrennan) [inline src/azul/drs.py:258]: I'm not necessarily objecting to it, but what exactly is gained by bundling all of the arguments into an object? The URL parsing is done here, but that could also be done in a separate method called within `get_object`.

File: src/azul/drs.py [hunk 2]
```diff
@@ -254,36 +254,49 @@ def access_url(url):
...
+    drs_uri: str
+    access_id: Optional[str] = None
+    access_method: AccessMethod = AccessMethod.https
+
+    @property
+    def url(self) -> str:
```
[the remark this answers is not in the document]
Comment (danielsotirhos) [reply src/azul/drs.py:264]: Hm. Thanks I'll consider that. The trick is the URL needs to be built possibly twice if first an `access_id` is not given and then obtained from the response.

File: src/azul/drs.py [hunk 3]
```diff
@@ -254,36 +254,49 @@ def access_url(url):
...
+        parsed = furl(self.drs_uri)
+        file_uuid = str(parsed.path).lstrip('/')
+        parsed.scheme = 'https'
```
[the remark this answers is not in the document]
Comment (danielsotirhos) [reply src/azul/drs.py:267]: Yes, great idea. I will add that in.

File: src/azul/drs.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/azul/drs.py:264]: Duh :facepalm:

File: src/azul/drs.py (hunk 1 above)
Comment (danielsotirhos) [reply src/azul/drs.py:258]: The `ClientObject` class idea was born from the need to extract the netloc and `file_uuid` from the `drs_uri` and then pass those values along with the `access_id` and `access_method` variables to the private methods `_get_object()` and `_get_object_access()`. When first approaching this I stored these as instance variables in the `Client` class itself, however it started to seem a bit messy to me to be using `get_object()` as a constructor and method to return a resulting value.
I'll experiment with this again and/or see if just passing the variables to the functions separately is really that bad.

Comment (jessebrennan) [reply src/azul/drs.py:258]: > see if just passing the variables to the functions separately

 that's more what I was imagining. I can see how storing them in the client would get messy.

File: src/azul/drs.py [hunk 4]
```diff
@@ -269,21 +266,30 @@ def get_object(self,
...
-        url = self.url / object_id
+            return self._get_object_access(drs_uri, access_id, access_method)
+
+    def _get_url(self, drs_uri: str, access_id: Optional[str] = None) -> str:
+        # Note the desire to include any query params found in the DRS URI
+        # (eg 'drs://{netloc}/{id}?version=123') when building the URL.
+        parsed = furl(drs_uri)
+        assert parsed.scheme == 'drs', parsed.scheme
+        parsed.scheme = 'https'
+        file_uuid = str(parsed.path).lstrip('/')
```
Comment (jessebrennan) [inline src/azul/drs.py:279]: Nitpick: while "we all know" this is a file UUID, since we're in DRS land here I think it makes more sense to call this `drs_id` or `object_id` or something like that.

File: src/azul/drs.py [hunk 5]
```diff
@@ -269,21 +266,30 @@ def get_object(self,
...
-            return self._get_object(object_id, access_method=access_method)
+            return self._get_object(drs_uri, access_method)
         else:
-            return self._get_object_access(object_id, access_id, access_method=access_method)
-
-    def _get_object(self, object_id: str, access_method: AccessMethod) -> str:
-        url = self.url / object_id
+            return self._get_object_access(drs_uri, access_id, access_method)
+
+    def _get_url(self, drs_uri: str, access_id: Optional[str] = None) -> str:
```
Comment (jessebrennan) [inline src/azul/drs.py:273]: I was just reading the DRS spec, and I realized that even with your improvements, we're still only support one type of DRS URI---we're not supporting https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_compact_identifier_based_drs_uris.

I think we should leave this unsupported for now, but we _may_ want to add a comment or error message if the compact identifier based URIs are used.

Feel free to leave this out though, since our DRS client is just for internal use anyway.

File: src/azul/drs.py [hunk 6]
```diff
@@ -269,21 +266,30 @@ def get_object(self,
         depends on the access method specified.
         """
         if access_id is None:
```
Comment (jessebrennan) [inline src/azul/drs.py:268]: Consider calling `_get_url` here and pass the parsed URL instead. This would save 1 line of code :)

Comment (jessebrennan) [reply src/azul/drs.py:268]: To be totally clear, I mean calling `_get_url` before the `if`-`else`

Comment (danielsotirhos) [reply src/azul/drs.py:270]: Unfortunately both `_get_object()` and `_get_object_access()` need to call `_get_url()` individually since `_get_object()` can obtain an `access_id` and then call `_get_object_access()` with it which requires a new URL being generated.

**Comment by danielsotirhos** [comment]
Retracting review request to fix an issue.

File: src/azul/drs.py (hunk 6 above)
Comment (jessebrennan) [reply src/azul/drs.py:270]: Good catch.

**Comment by danielsotirhos** [comment]
Taking ticket out of review to rebase on latest develop.

File: src/azul/drs.py [hunk 7]
```diff
@@ -269,21 +268,38 @@ def get_object(self,
...
-            return self._get_object(object_id, access_method=access_method)
+            return self._get_object(drs_uri, access_method)
         else:
-            return self._get_object_access(object_id, access_id, access_method=access_method)
+            return self._get_object_access(drs_uri, access_id, access_method)
 
-    def _get_object(self, object_id: str, access_method: AccessMethod) -> str:
-        url = self.url / object_id
+    @classmethod
```
Comment (hannes-ucsc) [inline src/azul/drs.py:275]: Does this have to be a class method?

File: src/azul/drs.py [hunk 8]
```diff
@@ -256,11 +258,8 @@ def access_url(url):
 
 class Client:
```
Comment (hannes-ucsc) [inline src/azul/drs.py:259]:
```suggestion
class DRSClient:
```

File: test/integration_test.py [hunk 9]
```diff
@@ -328,12 +327,12 @@ def __check_manifest(self, file: IO[bytes], uuid_field_name: str) -> List[Mappin
         return rows
 
     def _test_drs(self, file_uuid: str):
-        base_url = config.service_endpoint() + http_object_path('')
-        client = Client(base_url)
+        client = Client()
```
Comment (hannes-ucsc) [inline test/integration_test.py:330]:
```suggestion
        drs = DRSClient()
```

File: src/azul/drs.py (hunk 7 above)
Comment (hannes-ucsc) [reply src/azul/drs.py:275]: And be public?

Comment (hannes-ucsc) [reply src/azul/drs.py:275]: If the motivation is in the chained PR, make it a protected instance method in this PR and then make the necessary modification in the other PR.

File: src/azul/drs.py [hunk 10]
```diff
@@ -269,21 +268,38 @@ def get_object(self,
...
+            return self._get_object_access(drs_uri, access_id, access_method)
 
-    def _get_object(self, object_id: str, access_method: AccessMethod) -> str:
-        url = self.url / object_id
+    @classmethod
+    def drs_uri_to_url(cls, drs_uri: str, access_id: Optional[str] = None) -> str:
+        """
+        Translate a DRS URI into a DRS URL. All query params included in the DRS
+        URI (eg '{drs_uri}?version=123') will be carried over to the DRS URL.
+        Only Hostname-based DRS URIs (drs://<hostname>/<id>) are supported and
```
Comment (hannes-ucsc) [inline src/azul/drs.py:280]:
```suggestion
        Only hostname-based DRS URIs (drs://<hostname>/<id>) are supported while
```

File: src/azul/drs.py [hunk 11]
```diff
@@ -269,21 +268,38 @@ def get_object(self,
...
+        not Compact Identifier-based URIs (drs://[provider_code/]namespace:accession)
```
Comment (hannes-ucsc) [inline src/azul/drs.py:281]:
```suggestion
        compact, identifier-based URIs (drs://[provider_code/]namespace:accession) are not.
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/jobs/5704

**Comment by hannes-ucsc** [comment]
Please also remove these commits from your branch:

* Temporarily double queue waiting time (#2147, #2189)
* Match Gitlab job timeout against that of the queue waiter
* Log # of expected bundles when waiting for queues

They landed as part of #2094.

File: src/azul/drs.py [hunk 12]
```diff
@@ -269,21 +268,38 @@ def get_object(self,
...
+        compact, identifier-based URIs (drs://[provider_code/]namespace:accession)
+        are not.
+        """
+        parsed = furl(drs_uri)
+        require(parsed.scheme == 'drs',
+                f'The DRS URI {drs_uri} does not have the correct scheme')
+        reject(':' in parsed.netloc or ':' in str(parsed.path),
+               f'The value {drs_uri} is not a valid Hostname-based DRS URI')
+        parsed.scheme = 'https'
+        object_id = str(parsed.path).lstrip('/')
```
Comment (hannes-ucsc) [inline src/azul/drs.py:289]: I think you should assert that the path has only one element. Then you can avoid the string conversion and the weird lstrip.

File: src/azul/drs.py [hunk 13]
```diff
@@ -269,21 +268,38 @@ def get_object(self,
...
+        Translate a DRS URI into a DRS URL. All query params included in the DRS
+        URI (eg '{drs_uri}?version=123') will be carried over to the DRS URL.
+        Only hostname-based DRS URIs (drs://<hostname>/<id>) are supported while
+        compact, identifier-based URIs (drs://[provider_code/]namespace:accession)
+        are not.
+        """
+        parsed = furl(drs_uri)
+        require(parsed.scheme == 'drs',
+                f'The DRS URI {drs_uri} does not have the correct scheme')
+        reject(':' in parsed.netloc or ':' in str(parsed.path),
```
Comment (hannes-ucsc) [inline src/azul/drs.py:286]: Why does a colon in netloc indicate the URI is not hostname based?

Same question for path.

File: src/azul/drs.py [hunk 14]
```diff
@@ -269,21 +268,38 @@ def get_object(self,
...
+               f'The value {drs_uri} is not a valid Hostname-based DRS URI')
```
Comment (hannes-ucsc) [inline src/azul/drs.py:287]:
```suggestion
               f'The value {drs_uri} is not a valid hostname-based DRS URI')
```

When I requested the corresponding change to the docstring, you should have picked this up on your own.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/drs.py (hunk 13 above)
Comment (danielsotirhos) [reply src/azul/drs.py:286]: From: https://ga4gh.github.io/data-repository-service-schemas/preview/develop/docs/#_for_drs_clients
"The colon character is not allowed in a Hostname-based DRS URI, making it easy to tell them apart."

Hostname-based URI: `drs://<hostname>/<id>`
Compact Identifier-based URI: `drs://[provider_code/]namespace:accession`

The reason both `netloc` and `path` need to be checked for a colon is because the `provider_code` and following slash is optional in CI-based URI, which means when parsed by `furl` the colon could end up in the `netloc` or `path` depending if a `provider_code` was included or not.

Comment (hannes-ucsc) [reply src/azul/drs.py:286]: That's actually an issue with the DRS standard, making it hard to parse when following RFC3986.

https://tools.ietf.org/html/rfc3986#section-3

In `drs://foo:bar` (the variant without provider code), `foo:bar` matches the `hier-part` production of the RFC grammar.

In `drs://fubaz/foo:bar` (the variant with provider code), `foo:bar` matches the `path` production. Both are valid URIs but the fact that the `namespace` and `accession` URI components end up matching different production in the RFC grammar, depending on the case, is problematic. We may want to raise this with the author if the standard. A better alternative standard would be 

`drs://[provider_code]/namespace:accession`

**Review by hannes-ucsc — COMMENTED** [review]
Let's discuss after stand-up.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Not much to demo here, will be demoed as part of PR for #1764.
